File: solutions/cpp/Power_set.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cmath>
#include <iostream>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

// @include
void GeneratePowerSet(const vector<int>& S) {
  for (int i = 0; i < (1 << S.size()); ++i) {
    int x = i;
    while (x) {
      int tar = log2(x & ~(x - 1));
      cout << S[tar];
      if (x &= x - 1) {
        cout << ',';
      }
    }
    cout << endl;
  }
}
// @exclude

int main(int argc, char* argv[]) {
  vector<int> S;
  if (argc >= 2) {
    for (int i = 1; i < argc; ++i) {
      S.emplace_back(atoi(argv[i]));
    }
  } else {
    default_random_engine gen((random_device())());
    uniform_int_distribution<int> dis(1, 10);
    S.resize(dis(gen));
    for (int i = 0; i < S.size(); ++i) {
      S[i] = i;
    }
  }
  GeneratePowerSet(S);
  return 0;
}

File: solutions/python/power_set.py
# Power_set.cpp 848813e190b1b85a8e75107fe8513c3be38ad1a9
import sys
import math
import random


# @include
def generate_power_set(S):
    for i in range(1 << len(S)):
        x = i
        while x:
            tar = int(math.log2(x & ~(x - 1)))
            print(S[tar], end='')
            x &= x - 1
            if x:
                print(',', end='')
        print()
# @exclude


def main():
    if len(sys.argv) >= 2:
        S = [int(i) for i in sys.argv[1:]]
    else:
        S = list(range(random.randint(1, 10)))
    generate_power_set(S)


if __name__ == '__main__':
    main()
